WordPress/src/main/java/org/wordpress/android/WordPress.java

```diff
@@ -523,13 +523,26 @@ public void onAccountChanged(OnAccountChanged event) {
                 appLock.setPassword(null);
             }
         }
-        if (!event.isError() && event.causeOfChange == AccountAction.FETCH_SETTINGS && mAccountStore.hasAccessToken()) {
-            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
-            boolean hasUserOptedOut = !prefs.getBoolean(getString(R.string.pref_key_send_usage), true);
-            AnalyticsTracker.setHasUserOptedOut(hasUserOptedOut);
-            // When local and remote prefs are different, force opt out to TRUE
-            if (hasUserOptedOut != mAccountStore.getAccount().getTracksOptOut()) {
-                AnalyticsUtils.updateAnalyticsPreference(getContext(), mDispatcher, mAccountStore, true);
+
+        if (!event.isError() && mAccountStore.hasAccessToken()) {
+            // previously we reset the reader database on logout but this meant losing saved posts
+            // so now we only reset it when the user id changes
+            if (event.causeOfChange == AccountAction.FETCH_ACCOUNT) {
+                long thisUserId = mAccountStore.getAccount().getUserId();
+                long lastUserId = AppPrefs.getLastUsedUserId();
```

Comment:
Thanks @nbradbury! Unfortunately I don't have enough time to do full review/test today. However I was curious how you tackled this issue and I'm not sure about one thing. Won't the `AppPrefs.getLastUsedUserId()` by 0 for all users who are already logged in and who just update the app? Wdyt?

Reply:
The db would be reset for these people but I figured that would be okay since the "save for later" db version change would cause it to be reset anyway. I can check for zero, though, if you feel it's necessary.

Reply:
The `AppPrefs.setLastUsedUserId(thisUserId)` is called after a user logs in right? Is it called after the application is updated as well?
Scenario
1. User logs in with WP 9.7
2. User updates the app to WP 10.1 (assuming 10.1 includes saved post feature)
3. The ReaderDatabase is reset, but it doesn't matter as you mentioned
4. User saves some posts
5. User logs out -> isn't the `AppPrefs.getLastUsedUserId()` still 0 even though the user has already saved some posts?

If it's not the case or I haven't explained it enough, just forget about it and I'll test it tomorrow;). Thanks

Reply:
> User logs out -> isn't the AppPrefs.getLastUsedUserId() still 0 even though the user has already saved some posts?

Nope. The `onAccountChanged()` method is somewhat misnamed because it actually happens every time the app starts (when the account is fetched). So, the first time the user starts the app their userId would be saved. This method is also triggered when the user signs in.

Reply:
Ohh I see. Thanks for the explanation;)!